core/src/main/java/jenkins/model/GlobalToolConfiguration.java

```diff
@@ -0,0 +1,102 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.model;
+
+import com.google.common.base.Predicate;
+import hudson.Extension;
+import hudson.Functions;
+import hudson.model.Descriptor;
+import hudson.model.ManagementLink;
+import hudson.security.Permission;
+import hudson.util.FormApply;
+import net.sf.json.JSONObject;
+import org.kohsuke.stapler.StaplerRequest;
+import org.kohsuke.stapler.StaplerResponse;
+
+import javax.servlet.ServletException;
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+@Extension(ordinal = Integer.MAX_VALUE - 220)
+public class GlobalToolConfiguration extends ManagementLink {
+
+    @Override
+    public String getIconFileName() {
+        return "setting.png";
+    }
+
+    @Override
+    public String getDisplayName() {
+        return jenkins.management.Messages.ConfigureTools_DisplayName();
+    }
+
+    @Override
+    public String getDescription() {
+        return jenkins.management.Messages.ConfigureTools_Description();
+    }
+
+    @Override
+    public String getUrlName() {
+        return "configureTools";
+    }
+
+    @Override
+    public Permission getRequiredPermission() {
+        return Jenkins.ADMINISTER;
+    }
+
+    public synchronized void doConfigure(StaplerRequest req, StaplerResponse rsp) throws IOException, ServletException, Descriptor.FormException {
+        boolean result = configure(req, req.getSubmittedForm());
+        LOGGER.log(Level.FINE, "tools saved: "+result);
+        FormApply.success(req.getContextPath() + "/manage").generateResponse(req, rsp, null);
+    }
+
+    private boolean configure(StaplerRequest req, JSONObject json) throws hudson.model.Descriptor.FormException {
+        Jenkins j = Jenkins.getInstance();
+        j.checkPermission(Jenkins.ADMINISTER);
+
+        boolean result = true;
+        for(Descriptor<?> d : Functions.getSortedDescriptorsForGlobalConfig(FILTER)){
+            result &= configureDescriptor(req, json, d);
+        }
+
+        return result;
+    }
+
+    private boolean configureDescriptor(StaplerRequest req, JSONObject json, Descriptor<?> d) throws Descriptor.FormException {
+        String name = d.getJsonSafeClassName();
+        JSONObject js = json.has(name) ? json.getJSONObject(name) : new JSONObject(); // if it doesn't have the property, the method returns invalid null object.
+        json.putAll(js);
+        return d.configure(req, js);
```

Comment:
This seems like rather low-level form binding code that ought to live in a more generic place.